rapids-cmake/cython/add_rpath_entries.cmake

```diff
@@ -0,0 +1,90 @@
+# =============================================================================
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied. See the License for the specific language governing permissions and limitations under
+# the License.
+# =============================================================================
+
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_cython_add_rpath_entries
+-------------------------------
+
+.. versionadded:: v22.12.00
+
+Set the RPATH entries for all targets associated with a provided associated target.
+
+.. code-block:: cmake
+
+rapids_cython_add_rpath_entries(
+    TARGET <associated_target>
+    PATHS <path1> <path2> ...
+    [ROOT_DIRECTORY <root-dir>]
+)
+
+This function will affect all targets created up to the point of this call. It
+will have no effect on targets created afterwards.
+
+``TARGET``
+  The associated target for which we are setting RPATH entries. Any target
+  created using :code:command:`rapids_cython_create_modules` with the argument
+  `ASSOCIATED_TARGET associated_target` will have its RPATH entries updated.
+
+``PATHS``
+  The paths to add to the RPATH. Paths may either be absolute or relative to
+  the ROOT_DIRECTORY. The paths are always converted to be relative to the
+  current directory i.e relative to $ORIGIN in the RPATH.
+
+``ROOT_DIRECTORY``
+  The ROOT_DIRECTORY for the provided paths. Defaults to ${PROJECT_SOURCE_DIR}.
+  Has no effect on absolute paths.
+
+#]=======================================================================]
+function(rapids_cython_add_rpath_entries)
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.cython.add_rpath_entries")
+
+  set(options)
+  set(one_value ROOT_DIRECTORY TARGET)
+  set(multi_value PATHS)
+  cmake_parse_arguments(_RAPIDS_CYTHON "${options}" "${one_value}" "${multi_value}" ${ARGN})
+
+  # By default paths are relative to the current project root.
+  if(NOT _RAPIDS_CYTHON_ROOT_DIRECTORY)
+    set(_RAPIDS_CYTHON_ROOT_DIRECTORY "${PROJECT_SOURCE_DIR}")
+  endif()
+
+  # Transform all paths to paths relative to the current directory.
+  set(cleaned_paths)
+  foreach(path IN LISTS _RAPIDS_CYTHON_PATHS)
+    if(NOT IS_ABSOLUTE path)
+      # Convert relative paths to paths from the provided root, then make them relative to the
+      # current directory.
+      cmake_path(ABSOLUTE_PATH path BASE_DIRECTORY "${_RAPIDS_CYTHON_ROOT_DIRECTORY}")
+      cmake_path(ABSOLUTE_PATH path)
```

Comment:
I suspect that I am doing something wrong with the previous call. Currently this call is necessary to handle the case when a root directory is passed. For some reason I get a relative path back if `BASE_DIRECTORY != PROJECT_SOURCE_DIR`. If I add a `message("The path is ${path}")` before and after each of these `cmake_path` calls, here is what I currently get when [this test](https://github.com/rapidsai/rapids-cmake/pull/295/files#diff-ce676ffc31fdefa61b46bbf0206a8676379dd1a2faad4efd3472bc9be05babd7R48) runs:
```
538: The path is ../libraries
538: The path is ../../libraries
538: The path is /nvme/0/vyasr/rapids/rapids-cmake/testing/cython/add_rpath_entries/../../libraries
```

If I remove the second `cmake(ABSOLUTE_PATH)`, I instead get:
```
538: The path is ../libraries
538: The path is ../../libraries
538: The path is ../../libraries
```
Evidently the first call is in fact _not_ producing an absolute path in this scenario. Note that [the first test (where I don't provide a `BASE_DIRECTORY`)](https://github.com/rapidsai/rapids-cmake/pull/295/files#diff-ce676ffc31fdefa61b46bbf0206a8676379dd1a2faad4efd3472bc9be05babd7R41) does not exhibit this problem. Am I using `cmake_path` incorrectly in some way here?

Reply:
The issue is that `_RAPIDS_CYTHON_ROOT_DIRECTORY` isn't an absolute directory path. So when you call `cmake_path` the first time it can't make it a true absolute path as it has no idea on what directory you want to evaluate the relative `BASE_DIRECTORY` in regards too. The second one gets you a proper absolute path as now we evaluate based on `CMAKE_CURRENT_SOURCE_DIR`